src/python_testing/TC_SC_3_6.py

```diff
@@ -0,0 +1,174 @@
+                for controller in new_controllers:
+                    controller.name = all_names.pop(0)
+
+                client_list.extend(new_controllers)
+
+        asserts.assert_equal(len(client_list), num_fabrics_to_commission *
+                             num_controllers_per_fabric, "Must have the right number of clients")
+
+        # Before subscribing, set the NodeLabel to "Before Subscriptions"
+        logging.info("Pre-conditions: writing initial value of NodeLabel, so that we can control for change of attribute detection")
+        await client_list[0].WriteAttribute(self.dut_node_id, [(0, Clusters.Basic.Attributes.NodeLabel(value=BEFORE_LABEL))])
+
+        # Subscribe with all clients to NodeLabel attribute
+        subscriptions = []
+        sub_handlers = []
+        resub_catchers = []
+
+        logging.info("Step 1 (first part): Establish subscription with all %d clients" % len(client_list))
+        for client in client_list:
+            logging.info("Establishing subscription from controller node %s" % client.name)
+            sub = await client.ReadAttribute(nodeid=self.dut_node_id, attributes=[(0, Clusters.Basic.Attributes.NodeLabel)],
+                                             reportInterval=(min_report_interval_sec, max_report_interval_sec), keepSubscriptions=False)
+            subscriptions.append(sub)
+
+            resub_catcher = ResubscriptionCatcher(name=client.name)
+            sub.SetResubscriptionAttemptedCallback(resub_catcher)
+            resub_catchers.append(resub_catcher)
+
+            if sub_liveness_override_ms is not None:
+                logging.warning("Overriding subscription liveness to check %dms! NOT FOR CERTIFICATION!" % sub_liveness_override_ms)
+                sub.OverrideLivenessTimeoutMs(sub_liveness_override_ms)
+
+        asserts.assert_equal(len(subscriptions), len(client_list), "Must have the right number of subscriptions")
+
+        # Trigger a change on NodeLabel
+        logging.info(
+            "Step 1 (second part): Change attribute with one client, ensure no resubscriptions are triggered due to CASE session starvation")
+        await asyncio.sleep(1)
+        await client_list[0].WriteAttribute(self.dut_node_id, [(0, Clusters.Basic.Attributes.NodeLabel(value=AFTER_LABEL))])
+
+        # Await a stabilization delay in increments to let the event loops run
+        start_time = time.time()
+        while (time.time() - start_time) < stabilization_delay_sec:
+            await asyncio.sleep(0.05)
+
+        # After stabilization, validate no resubscriptions
+
+        logging.info("Validation of results")
+        failed = False
+
+        for catcher in resub_catchers:
+            if catcher.caught_resubscription:
+                logging.error("Client %s saw a resubscription" % catcher.name)
```

Comment:
One of the biggest issues I have with our current Python CI tests is that they print out errors like this (sometimes multiple times with the same exact print), making it very difficult to know which code-path failed.

I think it'd be much better if we just directly called `asserts.fail` directly here instead.

That brings up the question then of 'clean-up'. Do we still have an expectation that we want the next set of tests to be able to run after a failed test here? I'd think the answer there is no, since you cannot assume anything about the state on the target anymore. If so, then I think asserting here directly is fine...

Reply:
Did more clean-up. Keeping the final report due to otherwise not sure the exception will be in the right thread